views/js/runner/plugins/navigation/next/nextItemWarning.js

```diff
@@ -0,0 +1,155 @@
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *
+ * Copyright (c) 2019 (original work) Open Assessment Technologies SA ;
+ */
+
+/**
+ * Test Runner Navigation Plugin : nextItemWarning
+ *
+ * @author Martin Nicholson <martin@taotesting.com>
+ */
+define([
+    'jquery',
+    'lodash',
+    'i18n',
+    'taoTests/runner/plugin',
+    'taoQtiTest/runner/helpers/currentItem',
+    'taoQtiTest/runner/plugins/navigation/next/dialogConfirmNext'
+], function ($, _, __, pluginFactory, currentItemHelper, dialogConfirmNext){
+    'use strict';
+
+    /**
+     * Returns the configured plugin
+     */
+    return pluginFactory({
+        name : 'nextItemWarning',
+
+        /**
+         * Initialize the plugin (called during runner's init)
+         */
+        init : function init(){
+            var self = this;
+            var testRunner = this.getTestRunner();
+            var testData = testRunner.getTestData();
+            var testConfig = testData.config || {};
+            var testStore = testRunner.getTestStore(); // we'll store user's checkbox choice in here
+            testStore.setVolatile(self.getName());
+
+            /**
+             * Provides different variants of message text
+             * @returns {String}
+             */
+            function getCustomNextMessage(action) {
+                var customNextMessage;
+                var itemPartiallyAnswered = currentItemHelper.isAnswered(testRunner, true);
+                if (! itemPartiallyAnswered) {
+                    customNextMessage = __('Are you sure you want to go to the next item? You will not be able to go back and provide an answer.');
+                }
+                else if (action === 'next') {
+                    customNextMessage = __('Are you sure you want to go to the next item? You will not be able to go back and change your answer.');
+                }
+                else if (action === 'skip') {
+                    customNextMessage = __('Are you sure you want to clear your answer and go to the next item? You will not be able to go back and provide an answer.');
+                }
+                return customNextMessage;
+            }
+
+            //plugin behavior
+            /**
+             * Checks configuration, shows a dialog asking to confirm the nav action
+             *
+             * @param {String} action - 'next' or 'skip'
+             * @returns {Promise} - resolves if dialog accepted or not shown, rejects if dialog cancelled
+             */
+            function doNextWarning(action) {
+                //var context = testRunner.getTestContext();
+                var checkboxParams = null;
+
+                testRunner.trigger('disablenav');
+
+                return new Promise(function(resolve, reject) {
```

Comment:
Please chain promises : 
```js
return testStore.getStore(self.getName()).then(function(store) {
    return store.getItem('dontShowNextItemWarning').then(function(checkboxValue) {
        // Show the warning unless user has turned it off:
        if (checkboxValue !== true) {
            // Define checkbox only if enabled by config:
            if (testConfig.enableNextItemWarningCheckbox) {
                checkboxParams = {
                    checked: checkboxValue,
                    submitChecked: function() {
                        store.setItem('dontShowNextItemWarning', true);
                    },
                    submitUnchecked: function() {
                        store.setItem('dontShowNextItemWarning', false);
                    },
                };
                // use a wrapping promise only when necessary, since you're already in a chain,
                // 
                return new Promise(function(resolve, reject) {
                    dialogConfirmNext(
                        __('Go to the next item?'),
                        getCustomNextMessage(action),
                        resolve, // if the test taker accepts
                        function cancel() {
                           //this is an internal convention to differentiate errors, and rejection that cancel an aciton
                            reject({
                                cancel: true
                            }); 
                        },
                        checkboxParams
                    );
                });
            }
        }
    });
})
.catch(function(err) {
    //if the rejection is due to an error, please rethrow it
    if (err && err instanceof Error) {
        throw err;
    }
    if (err && err.cancel === true) {
        testRunner.trigger('enablenav');
    }
});

```

Reply:
Implemented following this model.

NB: I am still returning a rejected promise after catching the user's rejection, otherwise the move takes place (which is not desired).